handlers/jiochat/jiochat.go

```diff
@@ -0,0 +1,197 @@
+	err := handlers.DecodeAndValidateQueryParams(jcVerify, r)
+	if err != nil {
+		return nil, err
+	}
+
+	stringSlice := []string{channel.StringConfigForKey(courier.ConfigSecret, ""), jcVerify.Timestamp, jcVerify.Nonce}
+	sort.Sort(sort.StringSlice(stringSlice))
+
+	value := strings.Join(stringSlice, "")
+
+	hashObject := sha1.New()
+	hashObject.Write([]byte(value))
+	signatureCheck := hex.EncodeToString(hashObject.Sum(nil))
+
+	ResponseText := "unknown request"
+	StatusCode := 400
+
+	if signatureCheck == jcVerify.Signature {
+		ResponseText = jcVerify.EchoStr
+		StatusCode = 200
+	}
+
+	w.Header().Set("Content-Type", "text/plain")
+	w.WriteHeader(StatusCode)
+	_, err = fmt.Fprint(w, ResponseText)
+	return nil, err
+}
+
+type jiochatMsgRequest struct {
+	FromUsername string `json:"FromUserName"`
+	CreateTime   int64  `json:"CreateTime"`
+	MsgType      string `json:"MsgType"`
+	MsgID        int64  `json:"MsgId"`
+	Event        string `json:"Event"`
+	Content      string `json:"Content"`
+	MediaID      string `json:"MediaId"`
+}
+
+// ReceiveMessage is our HTTP handler function for incoming messages
+func (h *handler) ReceiveMessage(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	jcRequest := &jiochatMsgRequest{}
+	err := handlers.DecodeAndValidateJSON(jcRequest, r)
+	if err != nil {
+		return nil, err
+	}
+
+	if jcRequest.FromUsername == "" || jcRequest.MsgType == "" || (jcRequest.MsgID == 0 && jcRequest.Event == "") {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("missing parameters, must have 'FromUserName', 'MsgType' and either 'MsgId' or 'Event'"))
+	}
+
+	date := time.Unix(jcRequest.CreateTime, 0).UTC()
+
+	urn := urns.NewURNFromParts(urns.JiochatScheme, jcRequest.FromUsername, "")
+
+	if jcRequest.MsgType == "event" && jcRequest.Event == "subscribe" {
+
+		// build the channel event
+		channelEvent := h.Backend().NewChannelEvent(channel, courier.Follow, urn)
+
+		err := h.Backend().WriteChannelEvent(ctx, channelEvent)
+		if err != nil {
+			return nil, err
+		}
+
+		return []courier.Event{channelEvent}, courier.WriteChannelEventSuccess(ctx, w, r, channelEvent)
+	}
+
+	if jcRequest.MsgType == "event" {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("unknown event"))
+	}
+
+	// create our message
+	msg := h.Backend().NewIncomingMsg(channel, urn, jcRequest.Content).WithExternalID(fmt.Sprintf("%d", jcRequest.MsgID)).WithReceivedOn(date)
+	if jcRequest.MsgType == "image" || jcRequest.MsgType == "video" || jcRequest.MsgType == "voice" {
+		mediaURL := resolveFileID(jcRequest.MediaID)
+		msg.WithAttachment(mediaURL)
+	}
+
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
+}
+
+var mediaDownloadURL = "https://channels.jiochat.com/media/download.action"
+
+func resolveFileID(mediaID string) string {
```

Comment:
`resolveFileID` here is just building the URL however `msg.go` won't be able to `downloadAndSavetoS3` without access to jiochat channel access token, We need to discuss the approach to notify the authentication to msg.go